runtime/src/accounts.rs

```diff
@@ -1033,28 +1130,276 @@ mod tests {
+            0,
+            2,
+            vec![keypair1.pubkey(), keypair3.pubkey(), native_loader::id()],
+            Hash::default(),
+            instructions,
+        );
+        let tx1 = Transaction::new(&[&keypair1], message, Hash::default());
+        let txs = vec![tx0, tx1];
+        let results1 = accounts.lock_accounts(&txs);
+
+        assert!(results1[0].is_ok()); // Credit-only account (keypair1) can be referenced multiple times
+        assert!(results1[1].is_err()); // Credit-only account (keypair1) cannot also be locked as credit-debit
+        assert_eq!(
+            *accounts
+                .credit_only_account_locks
+                .read()
+                .unwrap()
+                .get(&keypair1.pubkey())
+                .unwrap()
+                .lock_count
+                .lock()
+                .unwrap(),
+            2
+        );
+
+        accounts.unlock_accounts(&[tx], &results0);
+        accounts.unlock_accounts(&txs, &results1);
+
+        let instructions = vec![CompiledInstruction::new(2, &(), vec![0, 1])];
+        let message = Message::new_with_compiled_instructions(
+            1,
+            0,
+            2,
+            vec![keypair1.pubkey(), keypair3.pubkey(), native_loader::id()],
+            Hash::default(),
+            instructions,
+        );
+        let tx = Transaction::new(&[&keypair1], message, Hash::default());
+        let results2 = accounts.lock_accounts(&[tx]);
+
+        assert!(results2[0].is_ok()); // Now keypair1 account can be locked as credit-debit
+
+        // Check that credit-only credits are still cached in accounts struct
+        let credit_only_account_locks = accounts.credit_only_account_locks.read().unwrap();
+        let keypair1_lock = credit_only_account_locks.get(&keypair1.pubkey());
+        assert!(keypair1_lock.is_some());
+        assert_eq!(*keypair1_lock.unwrap().lock_count.lock().unwrap(), 0);
+    }
+
+    #[test]
+    fn test_accounts_locks_multithreaded() {
+        let counter = Arc::new(AtomicU64::new(0));
+        let exit = Arc::new(AtomicBool::new(false));
+
+        let keypair0 = Keypair::new();
+        let keypair1 = Keypair::new();
+        let keypair2 = Keypair::new();
+
+        let account0 = Account::new(1, 0, &Pubkey::default());
+        let account1 = Account::new(2, 0, &Pubkey::default());
+        let account2 = Account::new(3, 0, &Pubkey::default());
+
+        let accounts = Accounts::new(None);
+        accounts.store_slow(0, &keypair0.pubkey(), &account0);
+        accounts.store_slow(0, &keypair1.pubkey(), &account1);
+        accounts.store_slow(0, &keypair2.pubkey(), &account2);
+
+        let accounts_arc = Arc::new(accounts);
+
+        let instructions = vec![CompiledInstruction::new(2, &(), vec![0, 1])];
+        let credit_only_message = Message::new_with_compiled_instructions(
+            1,
+            0,
+            2,
+            vec![keypair0.pubkey(), keypair1.pubkey(), native_loader::id()],
```

Comment:
If I'm reading this correctly, won't both this and the below transaction both mark keypair0.pubkey, keypair1.pubkey, and keypair2.pubkey as credit-only (because the number of credit-only accounts is marked as `2`) in both messages.

Reply:
No, you aren't reading it quite right. This tx is marking keypair1.pubkey and native_loader::id as credit-only, and the next is marking keypair2.pubkey and native_loader::id as credit-only. So keypair1.pubkey is credit-only in the first, and credit-debit in the second.

Reply:
ah I see, makes sense!